TM-9 Create Task Form Validation

TomsBemis opened the pull request:
**1) Clean create task form:**
<img width="286" alt="TM-9 PR Clean form" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/45bbdffa-76de-490c-89b7-6a5e777949fa">
**2) Title unique error:**
![image](https://github.com/TomsBemis/task-manager-frontend/assets/149372128/3618ca7e-cbe0-4d3a-a200-0c79a0dc6842)
**3) Touched create task form with errors:**
<img width="278" alt="TM-9 PR Touched form" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/f05380dd-c7ae-463b-9062-295b7f2ce03c">
**4) Valid create task form:**
<img width="281" alt="TM-9 PR Valid form" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/a3fa199f-455e-4d08-a542-e3d65e06ec66">

TomsBemis committed:
TM-9 Create task form validation

Attributes task type and status have enumerators. Task title validated to be unique and all attributes except description are required

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -1,22 +1,27 @@
+ 
   @Output() taskCreatedEvent = new EventEmitter<Task>();
 
+  taskTypes : {key: string, value: string}[] = [];
```
If you're going to use this type of object in multiple places it's better to create a new model for it

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -25,20 +30,35 @@ export class CreateTaskComponent {
+      });
+    });
+    // Map the task status enum to array of keys and values
+    Object.keys(TaskStatus).forEach(taskStatusKey => {
```
Object.entries(TaskStatus).forEach(([key, value]) => {
      this.taskStatuses.push({
        key : key, 
        value : value
      });
    });

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -25,20 +30,35 @@ export class CreateTaskComponent {
-  constructor(private taskService: TaskService) {}
+  constructor(private taskService: TaskService) {
+    // Map the task type enum to array of keys and values
+    Object.keys(TaskType).forEach(taskTypeKey => {
```
same here instead of going through the keys you can just go through the entries of the enum and have both key and value of that key

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -25,20 +30,35 @@ export class CreateTaskComponent {
       description: this.createTaskForm.get('description')?.value,
-      type: this.createTaskForm.get('type')?.value,
+      type: TaskType[this.createTaskForm.get('type')?.value as keyof typeof TaskType],
       createdOn: new Date(),
```
I noticed you have the date in the createForm why have it in the form if it's not being used?

wanpii commented on src/assets/tasks.json:
```diff
@@ -1,6 +1,6 @@
-    {"title" : "TM-4", "description" : "Task List Application - Create base", "type" : "Subtask", "createdOn" : "2024-05-23", "status": "Open"},
-    {"title" : "TM-5", "description" : "Create a “Create task“ component with add a task functionality using Angular Reactive Forms.", "type" : "Subtask", "createdOn" : "2024-05-23", "status": "In Progress"}
+    {"title" : "TM-2", "description" : "Setup Angular CLI", "type" : "STORY", "createdOn" : "2024-05-21", "status": "DONE"},
+    {"title" : "TM-3", "description" : "Create new application", "type" : "TASK", "createdOn" : "2024-05-22", "status": "OBSOLETE"},
```
I see that here you have the key as the value for type and status but when creating a new task you set the status and type as the value of the specific enum

wanpii replied:
see lines 59 and 61 in create task component

TomsBemis committed:
TM-9 Pull Request 3 requested changes

Field 'createdOn' removed from create task reactive form since timestamp is automatically generated. New data structure for task types and task statuses.

wanpii commented on src/app/tasks/create-task/create-task.component.html:
```diff
@@ -6,29 +6,57 @@
+                        <div class="col-sm-10">
+                            <input class="form-control" type="text" id="title" name="title" formControlName="title">
+                            @if (createTaskForm.get('title')?.errors) {
+                                @if (createTaskForm.get('title')?.errors!['required'] && createTaskForm.get('title')?.touched) {
```
non null assertion

TomsBemis replied:
Is non-null assertion missing or should I replace the optional operator with the non-null assertion? When I remove it from the 'errors' attribute I get an error, that it could be null.

wanpii replied:
![image](https://github.com/TomsBemis/task-manager-frontend/assets/107789468/80306328-f422-4053-93b4-acb304a97333)

wanpii commented on src/app/tasks/create-task/create-task.component.html:
```diff
@@ -6,29 +6,57 @@
+                                @if (createTaskForm.get('title')?.errors!['required'] && createTaskForm.get('title')?.touched) {
+                                    <span class="pull-right help-block">Title is required</span>
+                                }
+                                @if (createTaskForm.get('title')?.errors!['titleUnique']) {
```
same here

wanpii commented on src/app/tasks/task.service.ts:
```diff
@@ -14,14 +18,30 @@ export class TaskService {
-        taskData.forEach(taskElementData => {
+        // Load task types
+        taskData.taskTypes.forEach(taskType => {
+            this.taskTypes[taskType.value] = {
```
why use OptionIndex? You can just create an array of Option[] instead of indexing it, so taskTypes: Option[] = [] same for statuses, then you can just assign taskData.taskTypes to this.taskTypes

TomsBemis replied:
Because in TaskService I had to assign a specific TaskType/TaskStatus to the task, therefore I had "keys" to index the list of options.

wanpii replied:
You can use assign a specific type/status to a task without using indexing

wanpii commented on src/app/tasks/task.service.ts:
```diff
@@ -14,14 +18,30 @@ export class TaskService {
+        });
+
+        // Load task statuses
+        taskData.taskStatuses.forEach(taskStatus => {
```
same here

TomsBemis committed:
TM-9 Console log fix

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -1,44 +1,60 @@
 
+  @Input() taskTypes : Option[] = [];
+
+  @Input() taskStatuses : Option[] = [];
```
Yeah here you have Option array, where you can assign the types and statuses from the taskService

TomsBemis replied:
While in the TaskService I needed the Types and Statuses to be indexed so I can select a specific one, I needed to put them into arrays because I couldn't iterate over the indexed options in the template

wanpii replied:
You can use .find() method to find the one you're looking for

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -1,44 +1,60 @@
+  ngOnInit(): void {
+
+    // Map task types and statuses from index to arrays
+    Object.entries(this.taskService.getTaskTypes()).forEach(taskType => {
```
look at comment above

TomsBemis committed:
TM-9 Pull Request 3 requested changes #2

Removed option index and replaced with array of options

TomsBemis committed:
TM-9 Pull Request 3 Requested changes

Input properties assigned directly from service. Nullish coalescing operator used when finding task types and statuses in arrays.

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -1,31 +1,35 @@
+ 
   @Output() taskCreatedEvent = new EventEmitter<Task>();
 
+  @Input() taskTypes : Option[] = this.taskService.getTaskTypes();
```
I noticed that this is an input, but it is not used as an Input, if you don't plan on using it as an Input you can remove the @Input

TomsBemis committed:
TM-9 Pull Request 3 Requested changes: removed unnecessary input decorator
